go/vt/mysqlctl/backupengine.go

```diff
@@ -177,23 +187,38 @@ type BackupManifest struct {
 // FindBackupToRestore returns a selected candidate backup to be restored.
 // It returns the most recent backup that is complete, meaning it has a valid
 // MANIFEST file.
-func FindBackupToRestore(ctx context.Context, cnf *Mycnf, mysqld MysqlDaemon, logger logutil.Logger, dir string, bhs []backupstorage.BackupHandle) (backupstorage.BackupHandle, error) {
+func FindBackupToRestore(ctx context.Context, params RestoreParams, bhs []backupstorage.BackupHandle) (backupstorage.BackupHandle, error) {
 	var bh backupstorage.BackupHandle
 	var index int
+	// if a StartTime is provided in params, then find a backup that was taken before that time
+	checkBackupTime := !params.StartTime.IsZero()
 
 	for index = len(bhs) - 1; index >= 0; index-- {
 		bh = bhs[index]
 		// Check that the backup MANIFEST exists and can be successfully decoded.
-		_, err := GetBackupManifest(ctx, bh)
+		bm, err := GetBackupManifest(ctx, bh)
 		if err != nil {
-			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage: can't read MANIFEST: %v)", bh.Name(), dir, err)
+			params.Logger.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage: can't read MANIFEST: %v)", bh.Name(), params.Dir, err)
 			continue
 		}
 
-		logger.Infof("Restore: found backup %v %v to restore", bh.Directory(), bh.Name())
-		break
+		var backupTime time.Time
+		if checkBackupTime {
+			backupTime, err = time.Parse(time.RFC3339, bm.BackupTime)
+			if err != nil {
+				params.Logger.Warningf("Restore: skipping backup %v/%v with invalid time %v: %v", params.Dir, bh.Name(), bm.BackupTime, err)
+				continue
+			}
+		}
+		if !checkBackupTime /* not snapshot */ || backupTime.Before(params.StartTime) {
```

Comment:
super pedantic nit pick here -- we want to take a backup which was taken `<= StartTime` but this seems like it's going to be strictly `<`?

The main reason I'm calling this out here is that the process of moving over to use snapshot keyspaces will probably involve a phase where we don't have tooling in place yet to consume the actual restored time/position. As a result our PITR process will likely be:
1. find a specific backup through existing tooling
2. pass that into `CreateKeyspace --keyspace_type=SNAPSHOT` as our target time on the assumption it will find the specific backup we did in step 1

As written I believe that we would, instead, end up with the backup _one before_ what we would expect.

Reply:
This is a very good point. I had actually wanted to implement AtOrBefore semantics, and the only reason I didn't is because the time library doesn't actually have an AtOrBefore function (lame reason, I know). We can do `backupTime.Equal(..) || backupTime.Before(..)` instead. I will make this change.

Reply:
Done.